### PR TITLE
src/test: fix to avoid fail notification when testing manifest refcount

### PR DESCRIPTION
**Pull request by myoungwon**
Due to false-positive design on manifest snap refcounting,
a message to decrement the refcount can be missing.
This commit checks whether the manifest object's state is correct
when such mismatch happens to prevent aborting unit test.

The mismatch that happened here will be fixed by chunk scrub later.

The following is a specific scenario.
Assuming that there are osd 1, 6, 7, and object A is the manifest object in osd 1.

[osd 1, 6, 7], dec_refcount is invoked by osd 1 on object A

[osd 1, 6, 7], manifest info including chunk_info in object A is updated without waiting dec_refcount's completion

[osd 1 is out] a message to decrement the referece can not be delivered

[osd 6, 7] nothing happens, but the manifest info in the object A is already updated.

fixes: https://tracker.ceph.com/issues/48786
         https://tracker.ceph.com/issues/48915
         https://tracker.ceph.com/issues/47024

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by myoungwon** [comment]
Before going ahead, https://github.com/ceph/ceph/pull/38767 should be merged first.

**Comment by athanatos** [comment]
This appears to fix a problem introduced in #38767, why not include this in that PR?

**Comment by myoungwon** [comment]
I opend https://github.com/ceph/ceph/pull/38767 for appropriate refcounting after reviewing current chunk scrub code. Then, this issue is reported. So, I created another PR. Do you think this pr needs to be merged into https://github.com/ceph/ceph/pull/38767?

**Comment by athanatos** [comment]
If it's independent, than this is fine.

File: src/test/librados/tier_cxx.cc [hunk 1]
```diff
@@ -4482,8 +4537,20 @@ TEST_F(LibRadosTwoPoolsPP, ManifestSnapIncCount) {
...
+    cache_ioctx.getxattr("chunk4", CHUNK_REFCOUNT_ATTR, t);
+    chunk_refs_t refs;
+    try {
+      auto iter = t.cbegin();
+      decode(refs, iter);
+    } catch (buffer::error& err) {
+      ASSERT_TRUE(0);
+    }
+    if (refs.count() != 1u) {
+      ASSERT_EQ(true, is_intented_refcount_state(ioctx, "foo", cache_ioctx, "chunk4"));
```
Comment (tchaikov) [inline src/test/librados/tier_cxx.cc:4551]: how about using `ASSERT_TRUE()`?

File: src/test/librados/tier_cxx.cc [hunk 2]
```diff
@@ -5392,7 +5459,10 @@ TEST_F(LibRadosTwoPoolsPP, ManifestFlushDupCount) {
     sha1_gen.Final(fingerprint);
     buf_to_hex(fingerprint, CEPH_CRYPTO_SHA1_DIGESTSIZE, p_str);
     tgt_oid = string(p_str);
-    ASSERT_EQ(-ENOENT, ioctx.getxattr(p_str, CHUNK_REFCOUNT_ATTR, t));
+    int r = cache_ioctx.getxattr(p_str, CHUNK_REFCOUNT_ATTR, t);
+    if (r != -ENOENT) {
+      ASSERT_EQ(true, is_intented_refcount_state(cache_ioctx, "foo", ioctx, p_str));
```
Comment (tchaikov) [inline src/test/librados/tier_cxx.cc:5464]: ditto.

**Comment by myoungwon** [comment]
@tchaikov I addressed your comment.

**Comment by myoungwon** [comment]
@tchaikov Sorry, the failure in your recent test suite (https://pulpito.ceph.com/kchai-2021-01-27_08:19:23-rados-wip-kefu-testing-2021-01-27-1353-distro-basic-smithi/5833001/) caused by this PR because I added the code mistakenly, which invokes exec() with null string. I've re-pushed the commit to fix this, so it would be good if you include this PR when your next test suite runs.

**Comment by tchaikov** [comment]
@myoungwon  ack. will rerun the failed tests with the latest change in this PR.

**Comment by neha-ojha** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-02_15:01:12-rados-wip-kefu-testing-2021-02-02-1939-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_08:10:00-rados-wip-kefu-testing-2021-02-03-1357-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-03_14:17:43-rados-wip-kefu-testing-2021-02-03-2051-distro-basic-smithi/

**Comment by myoungwon** [comment]
@athanatos What do you think?

File: src/test/librados/tier_cxx.cc [hunk 3]
```diff
@@ -3621,7 +3658,9 @@ TEST_F(LibRadosTwoPoolsPP, ManifestSnapRefcount) {
     } catch (buffer::error& err) {
       ASSERT_TRUE(0);
     }
-    ASSERT_EQ(1u, refs.count());
+    if (refs.count() != 1u) {
```
Comment (athanatos) [inline src/test/librados/tier_cxx.cc:3661]: Why these if checks?  Is refs supposed to have just 1?

Comment (myoungwon) [reply src/test/librados/tier_cxx.cc:3661]: Yeah, The expected ref is 1 here. If the ref is not 1, just check whether the reference of both source and target is correct or not by the following code.

Comment (athanatos) [reply src/test/librados/tier_cxx.cc:3661]: I'm really not following.  Why not just embed that check in is_intended_refcount_state?

**Comment by myoungwon** [comment]
@athanatos I added a commit. Is this what you pointed out?

**Comment by athanatos** [comment]
What, specifically, is causing the refcount decrement to fail?  I don't see anything in this unit test that should cause that behaviour, so I wonder whether this patch would actually cover a bug.

**Comment by myoungwon** [comment]
Assuming that there are osd 1, 6, 7, and object A is the manifest object in osd 1.

[osd 1, 6, 7], dec_refcount is invoked by osd 1 on object A

[osd 1, 6, 7], manifest info including chunk_info in object A is updated without waiting dec_refcount's completion

[osd 1 is out] a message to decrement the referece can not be delivered

[osd 6, 7] nothing happens, but the manifest info in the object A is already updated.

**Comment by athanatos** [comment]
Does this test actually mark osd 1 out?

**Comment by athanatos** [comment]
Oh, I suppose it's running with osd thrashing in the background in teuthology?

**Comment by myoungwon** [comment]
Yes. The case I mentioned occurs when it's running in teuthology with osd thrashing.

File: src/test/librados/tier_cxx.cc [hunk 4]
```diff
@@ -155,6 +155,47 @@ string get_fp_oid(string oid, std::string fp_algo = NULL)
...
+  {
+    bufferlist in, out;
+    encode(dst_oid, in);
+    src_refcount = src_ioctx.exec(src_oid, "cas", "references_chunk", in, out);
+    if (src_refcount == -ENOENT || src_refcount == -ENOLINK) {
+      src_refcount = 0;
+    }
+    ceph_assert(src_refcount >= 0);
+  }
+  if (src_refcount > dst_refcount) {
```
Comment (athanatos) [inline src/test/librados/tier_cxx.cc:193]: Isn't this backwards?  The design is that the refcount recorded on the target is always >= than the real number of references, right?  We always increment the refcount on the target before updating the source metadata and update the source metadata before decrementing the target refcount, right?

Comment (athanatos) [reply src/test/librados/tier_cxx.cc:193]: src_refcount should always match expected_refcount, right?  I think you should always read dst_refcount and src_refcount and assert:
1. src_refcount == expected_refcount
2. dst_refcount >= src_refcount

Comment (myoungwon) [reply src/test/librados/tier_cxx.cc:193]: src means here is the manifest object on the base tier, and dst means that the the chunked object on the low tier.
So, isn't dst_refcount == expected_refcount correct?

Comment (athanatos) [reply src/test/librados/tier_cxx.cc:193]: I don't think so, we specifically allow dst_refcount to be larger.

Comment (myoungwon) [reply src/test/librados/tier_cxx.cc]: Yeah, dst_refcount >= src_refcount is correct.
but expected_refcount here is to check whether dst_refcount is the refcount as our expectation or not.

Comment (myoungwon) [reply src/test/librados/tier_cxx.cc]: Oh, sorry, I misunderstood. assert:

    src_refcount == expected_refcount
    dst_refcount >= src_refcount

are correct. Can you take a look a commit I added?

**Comment by myoungwon** [comment]
@athanatos any other comments?

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
Looks good to me.  Has this gone through testing yet?

**Comment by myoungwon** [comment]
@tchaikov @athanatos The fail in the last test run (https://pulpito.ceph.com/kchai-2021-02-14_08:35:02-rados-wip-kefu-testing-2021-02-14-1248-distro-basic-smithi/5880425/) occurs during ManifestSnapRefcount2() because I didn't catch a point that snap_remove is done asynchronously by selfmanaged_snap_remove(). So, cls_cas_references_chunk() to count source object's refcount is delivered before trim_object() is executed. To resolve this, I added a commit, which allow expected_refcounts to have multiple values, which considers two cases; the snapshot is removed or not.

@tchaikov Can you please re-include this PR when your next test run?

**Comment by athanatos** [comment]
I'm worried that that kind of renders the test pointless -- it'll succeed even if the snap removal never actually happens or if the snap removal happens, but the refcount update is incorrect.  What if you returned an error if there are untrimmed snaps?  You could then loop until the snaps have been trimmed.  You can check the current OSDMap for each snap to check whether it still exists (get_osdmap()->in_removed_snaps_queue(info.pgid.pgid.pool(), oid.snap)).

**Comment by myoungwon** [comment]
@athanatos Done. Make sense?

**Review by athanatos — APPROVED** [review]
Seems good pending a suite run.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-02-24_02:29:54-rados-wip-kefu-testing-2021-02-23-2359-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-24_15:54:55-rados-wip-kefu-testing-2021-02-24-2229-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-02-25_04:36:04-rados-wip-kefu-testing-2021-02-24-2230-distro-basic-smithi/

**Pull request merged**